---
title: "Basic CELLxGENE workflow"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Wetlab Module}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction
  
This vignette demonstrates a basic LaminDB workflow using the public CELLxGENE instance.
[CZ CELLxGENE Discover](https://cellxgene.cziscience.com/) is a standardised collection of scRNA-seq datasets and LaminDB makes it easy to query and access data in this repository.
We will go through the steps of finding and downloading a dataset using **{laminr}**, performing some simple analysis using **{Seurat}** and saving the results in a LaminDB database.

# Before we start

Before we go begin, please take some time to check out the Getting Started vignette (`vignette("laminr", package = "laminr")`).
In particular, make sure you have run the commands in the "Initial Setup" section.

Once that is done, we can load the **{laminr}** library.

```{r library}
library(laminr)
```

# Connect to a LaminDB instance

We will start by connecting to the CELLxGENE instance.
This gives us an object we can use to interact with the database.

```{r connect}
cellxgene <- connect("laminlabs/cellxgene")
cellxgene
```

# Downloading a dataset

In Lamin, artifacts are objects that[;] contain information (single-cell data, images, data frames etc.) as well as associated metadata.
You can see what artifacts are available using the database object.

```{r list-artifacts}
cellxgene$Artifact$df(limit = 5)
```

This is useful, but it's not the nicest or easiest way to find a particular dataset.
Instead, we will use the Lamin Hub website to find the data we want to load.

1. Open a browser and go to https://lamin.ai/laminlabs/cellxgene
2. On the top toolbar, click the "Artifacts" tab
3. Use the search field and the filters to find a dataset you are interested in.
  - We use the "Suffix" filter to find `.h5ad` files and search for "renal cell carcinoma"
4. Select the entry for the dataset you want to load to open a page with more details
5. Click the copy button at the top right, this copies a command including the ID for the artifact

Once we have the artifact ID, we can load information about the artifact, similar to what we see on the website.
Notice that we use a slightly different command to what we copied from the website.

```{r get-artifact}
artifact <- cellxgene$Artifact$get("7dVluLROpalzEh8mNyxk")
artifact
```

So far we have only retrieved the metadata about this object.
To download the data itself we need to run another command.

```{r load-artifact}
adata <- artifact$load()
adata
```

This dataset has been stored as an [`AnnData`](https://anndata.readthedocs.io) object.
In the next sections we will convert it to a [`Seurat`](https://satijalab.org/seurat/) object and perform some simple analysis.

# Convert to Seurat

There are various approaches for converting between different single-cell objects, some of which are described in the [Interoperability chapter](https://www.sc-best-practices.org/introduction/interoperability.html) of the Single-cell Best Practices book.

Because we already have the data loaded in memory, the simplest option is to extract the information we need and create a new `Seurat` object.

```{r create-seurat}
seurat <- SeuratObject::CreateSeuratObject(
  counts = Matrix::t(adata$X),
  meta.data = adata$obs, 
)
seurat
```

# Analysis

We could perform any normal analysis using **{Seurat}** but as an example we will calculate marker genes for each of the annotated cell types.
To make things a bit quicker we only test the first 1000 genes but if you have a few minutes you can get results for all features.

```{r markers}
# Set cell identities to the provided cell type annotation
SeuratObject::Idents(seurat) <- "Cell_Type"
# Normalise the data
seurat <- Seurat::NormalizeData(seurat)
# Test for marker genes
markers <- Seurat::FindAllMarkers(
  seurat, features = SeuratObject::Features(seurat)[1:1000]
)
# The output is a data.frame
head(markers)
```

# Add results

- Add the results as an artifact
- Render and add report (?)
