---
title: "Example Workflow: CELLxGENE"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Example Workflow: CELLxGENE}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
# whether or not this code will be used to
# actually upload results to the LaminDB instance
# -> testuser1 is a test account that cannot upload results
submit_eval <- laminr:::.get_user_settings()$handle != "testuser1"
```

# Introduction
  
This vignette demonstrates a basic workflow for accessing, analysing, and saving single-cell RNA-seq data from [laminlabs/cellxgene](https://lamin.ai/laminlabs/cellxgene), a mirror of [CZ CELLxGENE Discover](https://cellxgene.cziscience.com/), using **{laminr}**.

# Before we start

Ensure that you are familiar with the Getting Started vignette (`vignette("laminr", package = "laminr")`).
In particular, make sure you have run the commands in the "Initial Setup" section.

Load the **{laminr}** library.

```{r library}
library(laminr)
```

# Connecting to LaminDB

For this tutorial, you will connect a default instance (where we will store results) and the CELLxGENE instance that hosts the existing datasets.

## Connect to the default instance

Start by connecting to your default LaminDB instance.
You can set set the default instance using the `lamin` CLI on the command line:

```shell
lamin connect <owner>/<name>
```

Once a default instance has been set, connect to it with **{laminr}**:

```{r connect-default}
db <- connect()
db
```

This gives you a `db` object that you can use to interact with the instance.

**Note** that only the default instance can create new records.
This tutorial assumes you have access to an instance where you have `write` permission.

## Track data provenance

Track the code that is run in this notebook.

```{r track, eval = submit_eval}
db$track("I8BlHXFXqZOG0000", path = "example_workflow.Rmd")
```

Tip: The ID should be obtained by running `db$track(path = "example_workflow.Rmd")` and copying the ID from the output.

## Connect to the CELLxGENE instance

You can connect to other instances by providing a slug to the `connect()` function.
Instances connected to in this way can be used to query data but cannot make any changes.
Further connect to the CELLxGENE instance:

```{r connect-cellxgene}
cellxgene <- connect("laminlabs/cellxgene")
cellxgene
```

# Downloading a dataset

You can explore the available artifacts, which contain the data and associated metadata, through the instance object.

```{r list-artifacts}
cellxgene$Artifact$df(limit = 5)
```

Alternatively, use the [Lamin Hub](https://lamin.ai/laminlabs/cellxgene) to find the artifact that you want to load.
You can get information about the artifact using the artifact ID.
Notice the slightly different command to what was copied from the Lamin Hub.

```{r get-artifact}
artifact <- cellxgene$Artifact$get("7dVluLROpalzEh8mNyxk")
artifact
```

So far we have only retrieved the metadata about this object.
To download the data itself we need to run another command.

```{r load-artifact}
adata <- artifact$load()
adata
```

This dataset has been stored as an [`AnnData`](https://anndata.readthedocs.io) object that can be converted to a [`Seurat`](https://satijalab.org/seurat/) object for analysis.

# Convert to Seurat

There are various approaches for converting between different single-cell objects, some of which are described in the [Interoperability chapter](https://www.sc-best-practices.org/introduction/interoperability.html) of the Single-cell Best Practices book.

Because the data is loaded in memory, the simplest option is to extract the required information and to create a new `Seurat` object.

```{r create-seurat}
seurat <- SeuratObject::CreateSeuratObject(
  counts = Matrix::t(adata$X),
  meta.data = adata$obs,
)
seurat
```

# Analysis

You can now perform any normal analysis using **{Seurat}**.
Here, marker genes are exemplarily calculated for each of the annotated cell types.

```{r markers}
# Set cell identities to the provided cell type annotation
SeuratObject::Idents(seurat) <- "Cell_Type"
# Normalise the data
seurat <- Seurat::NormalizeData(seurat)
# Test for marker genes
# Subsample to first 1000 genes to speed up processing
markers <- Seurat::FindAllMarkers(
  seurat,
  features = SeuratObject::Features(seurat)[1:1000]
)
# The output is a data.frame
head(markers)
```

# Saving the results to your instance

Save results to your LaminDB instance.

```{r save-results, eval = submit_eval}
seu_path <- tempfile(fileext = ".rds")
saveRDS(seurat, seu_path)

db$Artifact$from_df(
  markers,
  description = "Marker genes for renal cell carcinoma dataset"
)$save()

db$Artifact$from_path(
  seu_path,
  description = "Seurat object for renal cell carcinoma dataset"
)$save()
```

# Finish track

Finish your tracked run to write a timestamp.

```{r close, eval = submit_eval}
db$finish()
```

## Render and upload the notebook

You can render this notebook to HTML:

- In RStudio, click the "Knit" button
- From the command line, run:

  ```bash
  Rscript -e 'rmarkdown::render("example_workflow.Rmd")'
  ```

- Or use the `rmarkdown` package in R:

  ```r
  rmarkdown::render("example_workflow.Rmd")
  ```

And then save it to your LaminDB instance using the CLI:

```bash
lamin save example_workflow.Rmd
```
