---
title: "Introduction"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


```{r prepare, include = FALSE}
# Restore the current instance at the end of the vignette
current_instance <- laminr::get_current_lamin_instance()
withr::defer(laminr::lamin_connect(current_instance))

# Restore the current user at the end of the vignette
current_user <- laminr::get_current_lamin_user()
withr::defer(laminr::lamin_login(current_user))

# Disconnect from this instance at the end of the vignette
withr::defer(laminr::lamin_disconnect())
```

This vignette introduces the **{laminr}** workflow.

To learn more about LaminDB, see [docs.lamin.ai](https://docs.lamin.ai).


```{r library}
library(laminr)
```

# Setup

Install **{laminr}** from CRAN:

```r
install.packages("laminr", dependencies = TRUE)
```

Install the underlying Python packages LaminDB and Bionty:

```r
laminr::install_lamindb(extra_packages = c("bionty"))
```

Set the default LaminDB instance:

```r
laminr::lamin_connect("<owner>/<name>")
```

This instance acts as the default instance for everything that follows.
Any data and tracking information will be added to it.

If you don't have access to an instance, create a local test instance.

```{r init}
lamin_init_tmp(name = "laminr-intro", modules = c("bionty"))
```

```{r disconnect, include = FALSE}
# Disconnect from this instance at the end of the vignette
withr::defer(laminr::lamin_disconnect())
```

# Import

To start working with **{laminr}**, import the **lamindb** module:

```{r import-lamindb}
ln <- import_module("lamindb")
```

This is equivalent to `import lamindb as ln` in Python.

# Walkthrough

This section of the vignette reproduces the walkthrough from the LaminDB [Introduction guide](https://docs.lamin.ai/guide).
The equivalent **{laminr}** code is included here, for the related text see the associated links.

See https://docs.lamin.ai/guide#walkthrough.

## Transforms

See https://docs.lamin.ai/guide#transforms.

```{r transforms}
ln <- import_module("lamindb")
ln$track()

ln$Transform$df()

ln$Run$df()
```

## Artifacts

Artifacts are objects that bundle data and associated metadata.
An artifact can be any file or folder but is typically a dataset.

See https://docs.lamin.ai/guide#artifacts.

```{r artifacts}
df <- ln$core$datasets$small_dataset1(otype = "DataFrame", with_typo = TRUE)
df

artifact <- ln$Artifact$from_df(df, key = "my_datasets/rnaseq1.parquet")$save()
artifact$describe()

artifact$cache()

dataset <- artifact$open()
as.data.frame(dataset)

artifact$load()

artifact$view_lineage()

df_typo <- df
levels(df$perturbation) <- c("DMSO", "IFNG")
df["sample2", "perturbation"] <- "IFNG"
artifact <- ln$Artifact$from_df(df, key = "my_datasets/rnaseq1.parquet")$save()
artifact$versions$df()
```

## Labels

See https://docs.lamin.ai/guide#labels.

```{r labels}
bt <- import_module("bionty")

experiment_type <- ln$ULabel(name = "Experiment", is_type = TRUE)$save()
candidate_marker_experiment <- ln$ULabel(
  name = "Candidate marker experiment", type = experiment_type
)$save()

artifact$ulabels$add(candidate_marker_experiment)

cell_type <- bt$CellType$from_source(name = "effector T cell")$save()
artifact$cell_types$add(cell_type)

artifact$describe()
```

## Registries

See https://docs.lamin.ai/guide#registries.

```{r registries}
ln$ULabel$df()

ln$Artifact
```

## Query & search

See https://docs.lamin.ai/guide#query-search.

```{r query-search}
transform <- ln$Transform$get(key = "introduction.Rmd")

ln$Artifact$filter(key__startswith = "my_datasets/")$df()

artifacts <- ln$Artifact$filter(transform = transform)$all()

artifacts <- ln$Artifact$filter(
  transform__description__icontains = "intro", ulabels = candidate_marker_experiment
)$all()

ln$Transform$search("intro")$df()
ulabels <- ln$ULabel$lookup()
cell_types <- bt$CellType$lookup()
```

## Features

See https://docs.lamin.ai/guide#features.

```{r features}
ln$Feature(name = "temperature", dtype = "float")$save()

ln$Feature(name = "experiment", dtype = ln$ULabel)$save()

artifact$features$add_values(
  list("temperature" = 21.6, "experiment" = "Candidate marker experiment")
)

artifact$describe()

ln$Artifact$features$filter(experiment__contains = "marker experiment")$df()
```

# Key use cases

This section of reproduces the key use cases from the LaminDB [Introduction guide](https://docs.lamin.ai/guide).

See https://docs.lamin.ai/guide#key-use-cases.

## Understand data lineage

See https://docs.lamin.ai/guide#understand-data-lineage.

```{r understanding-lineage}
artifact$view_lineage()
transform$view_lineage()
```

```r
# Example only, not run
ln <- import_module("lamindb")
ln$track()
ln$finish()

# lamin load https://lamin.ai/laminlabs/lamindata/transform/13VINnFk89PE0004
```

## Curate datasets

See https://docs.lamin.ai/introduction#curate-datasets.

```{r curate-datasets}
perturbation_type <- ln$ULabel(name = "Perturbation", is_type = TRUE)$save()
ln$ULabel(name = "DMSO", type = perturbation_type)$save()
ln$ULabel(name = "IFNG", type = perturbation_type)$save()

# Load Python built ins to get access to dtypes
py_builtins <- reticulate::import_builtins()

schema <- ln$Schema(
  name = "My DataFrame schema",
  features = list(
    # NOTE: These have dtype=int in the original guide
    ln$Feature(name = "ENSG00000153563", dtype = py_builtins$float)$save(),
    ln$Feature(name = "ENSG00000010610", dtype = py_builtins$float)$save(),
    ln$Feature(name = "ENSG00000170458", dtype = py_builtins$float)$save(),
    ln$Feature(name = "perturbation", dtype = ln$ULabel)$save()
  )
)$save()

curator <- ln$curators$DataFrameCurator(df, schema)
artifact <- curator$save_artifact(key = "my_curated_dataset.parquet")
artifact$describe()
ln$Artifact$get(ulabels__name = "IFNG")

curator <- ln$curators$DataFrameCurator(df_typo, schema)
tryCatch(
  curator$validate(),
  error = function(err) {
    cat(conditionMessage(err))
  }
)
```

## Manage biological registries

See https://docs.lamin.ai/introduction#manage-biological-registries.

```{r manage-biological-registries}
cell_types <- bt$CellType$public()
cell_types
cell_types$search("gamma-delta T cell") |> head(2)

var_schema <- ln$Schema(
  name = "my_var_schema",
  itype = bt$Gene$ensembl_gene_id,
  dtype = py_builtins$float
)$save()
obs_schema <- ln$Schema(
  name = "my_obs_schema",
  features = list(
    ln$Feature(name = "perturbation", dtype = ln$ULabel)$save()
  )
)$save()
anndata_schema <- ln$Schema(
  name = "my_anndata_schema",
  otype = "AnnData",
  components = list("obs" = obs_schema, "var" = var_schema)
)$save()

library(anndata)
adata <- AnnData(
  df[c("ENSG00000153563", "ENSG00000010610", "ENSG00000170458")],
  obs = df[, "perturbation", drop = FALSE]
)
curator <- ln$curators$AnnDataCurator(adata, anndata_schema)
artifact <- curator$save_artifact(description = "my RNA-seq")
artifact$describe()

genes <- bt$Gene$filter(organism__name = "human")$lookup()
feature_sets <- ln$FeatureSet$filter(genes = genes$cd8a)$all()
ln$Artifact$filter(feature_sets__in = feature_sets)$df()

neuron <- bt$CellType$from_source(name = "neuron")$save()
new_cell_state <- bt$CellType(
  name = "my neuron cell state", description = "explains X"
)$save()
new_cell_state$parents$add(neuron)
new_cell_state$view_parents(distance = 2)
```

## Scale learning

See https://docs.lamin.ai/introduction#scale-learning.

```{r scale-learning}
df2 <- ln$core$datasets$small_dataset2(otype = "DataFrame")
adata <- AnnData(
  df2[c("ENSG00000153563", "ENSG00000010610", "ENSG00000004468")],
  obs = df2[, "perturbation", drop = FALSE]
)
curator <- ln$curators$AnnDataCurator(adata, anndata_schema)
artifact2 <- curator$save_artifact(key = "my_datasets/my_rnaseq2.h5ad")

collection <- ln$Collection(
  list(artifact, artifact2),
  key = "my-RNA-seq-collection"
)$save()
collection$describe()
collection$view_lineage()

collection$load()
collection$artifacts$all()
collection$artifacts$df()
```

# Other examples

This section contains some additional examples of using **{laminr}**.

## Download a dataset form another instance

This examples shows how to get an artifact from another instance ("laminlabs/cellxgene").

```{r get-artifact}
artifact <- ln$Artifact$using("laminlabs/cellxgene")$get("7dVluLROpalzEh8mNyxk")
artifact
```

<div class="alert alert-info" role="alert">
**Tip**

You can view detailed information about this dataset on LaminHub: https://lamin.ai/laminlabs/cellxgene/artifact/7dVluLROpalzEh8mNyxk.

You can search and query more CELLxGENE datasets here: https://lamin.ai/laminlabs/cellxgene/artifacts.
</div>

To download the dataset and load it into memory, run:

```{r load-artifact}
adata <- artifact$load()
adata
```

This artifact contains an [`AnnData`](https://anndata.readthedocs.io) object.

<div class="alert alert-info" role="alert">
**Tip**

If you prefer a path to a local file or folder, call `path <- artifact$cache()`.
</div>

## Work with the dataset

Once you have loaded a dataset you can perform any analysis with it as you would normally.
Here, marker genes are calculated for each of the provided cell type labels using [**{Seurat}**](https://satijalab.org/seurat/).

```{r create-seurat}
library(Seurat)

# Create a Seurat object
seurat_obj <- CreateSeuratObject(
  counts = as(Matrix::t(adata$X), "CsparseMatrix"),
  meta.data = adata$obs
)
# Add gene metadata
seurat_obj[["RNA"]] <- AddMetaData(
  GetAssay(seurat_obj), adata$var
)
# Set cell identities to the provided cell type annotation
Idents(seurat_obj) <- "cell_type"
# Normalise the data
seurat_obj <- NormalizeData(seurat_obj)
# Test for marker genes (the output is a data.frame)
markers <- FindAllMarkers(
  seurat_obj,
  features = Features(seurat_obj)[1:100] # Only test a few features for speed
)
# Display the marker genes
knitr::kable(markers)
# Plot the marker genes
DotPlot(seurat_obj, features = unique(markers$gene)) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, vjust = 0.5))
```

## Slice a TileDB-SOMA array store

When artifacts contain TileDB-SOMA array stores they can be opened and sliced using the [**{tiledbsoma}** package](https://single-cell-data.github.io/TileDB-SOMA/index.html).

```{r slice-tiledbsoma, eval = requireNamespace("tiledbsoma", quietly = TRUE)}
# Set some environment variables to avoid an issue with {tiledbsoma}
# https://github.com/chanzuckerberg/cellxgene-census/issues/1261
Sys.setenv(TILEDB_VFS_S3_REGION = "us-west-2")
Sys.setenv(AWS_DEFAULT_REGION = "us-west-2")
Sys.setenv(TILEDB_VFS_S3_NO_SIGN_REQUEST = "true")

# Define a filter to select specific cells
value_filter <- paste(
  "tissue == 'brain' &&",
  "cell_type %in% c('microglial cell', 'neuron') &&",
  "suspension_type == 'cell' &&",
  "assay == '10x 3\\' v3'"
)

# Get the artifact containing the CELLxGENE Census TileDB-SOMA store
census_artifact <- ln$Artifact$using("laminlabs/cellxgene")$get("FYMewVq5twKMDXVy0001")
# Open the SOMACollection
soma_collection <- census_artifact$open()
# Slice the store to get a SOMADataFrame containing metadata for the cells of interest
cell_metadata <- soma_collection$get("census_data")$get("homo_sapiens")$obs$read(value_filter = value_filter)
# Concatenate the results to an arrow::Table
cell_metadata <- cell_metadata$concat()
# Convert to a data.frame
cell_metadata <- cell_metadata$to_data_frame()

cell_metadata
```

## Save the results

Save results as new artifacts to the current LaminDB instance.

```{r save-results}
seurat_path <- tempfile(fileext = ".rds")
saveRDS(seurat_obj, seurat_path)

ln$Artifact$from_df(
  markers,
  key = "laminr-datasets/renal-cell-carcinoma-markers.parquet",
  description = "Marker genes for renal cell carcinoma dataset"
)$save()

ln$Artifact(
  seurat_path,
  key = "laminr-datasets/renal-cell-carcinoma.rds",
  description = "Seurat object for renal cell carcinoma dataset"
)$save()
```

# Finish tracking

Mark the analysis run as finished to create a time stamp and upload source code to the hub.

```{r finish}
ln$finish()
```

## Save a notebook report (not needed for `.R` scripts)

Save a run report of your notebook (`.Rmd` or `.qmd` file) to your instance:

1. Render the notebook to HTML

- In RStudio, click the "Knit" button
- **OR** From the command line, run:

  ```bash
  Rscript -e 'rmarkdown::render("introduction.Rmd")'
  ```

- **OR** Use the `rmarkdown` package in R:

  ```r
  rmarkdown::render("introduction.Rmd")
  ```

2. Save it to your LaminDB instance:

- Using the `lamin_save()` function in R:

```r
lamin_save("introduction.Rmd")
```

- **OR** Using the `lamin` CLI:

```bash
lamin save introduction.Rmd
```

# Design

See https://docs.lamin.ai/introduction#design for more information on the design of LaminDB.
