<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Example Workflow: CELLxGENE • laminr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Example Workflow: CELLxGENE">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">laminr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/laminr.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Introduction</h6></li>
    <li><a class="dropdown-item" href="../articles/example_workflow.html">Example Workflow</a></li>
    <li><a class="dropdown-item" href="../articles/architecture.html">Package Architecture</a></li>
    <li><a class="dropdown-item" href="../articles/development.html">Development Roadmap</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>API reference</h6></li>
    <li><a class="dropdown-item" href="../articles/module_core.html">Core module</a></li>
    <li><a class="dropdown-item" href="../articles/module_bionty.html">Bionty module</a></li>
    <li><a class="dropdown-item" href="../articles/module_wetlab.html">Wetlab module</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/laminlabs/laminr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Example Workflow: CELLxGENE</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/laminlabs/laminr/blob/main/vignettes/example_workflow.Rmd" class="external-link"><code>vignettes/example_workflow.Rmd</code></a></small>
      <div class="d-none name"><code>example_workflow.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette demonstrates a basic workflow for accessing and
analysing single-cell RNA-seq data from the CELLxGENE repository using
<strong>{laminr}</strong>. <a href="https://cellxgene.cziscience.com/" class="external-link">CZ CELLxGENE Discover</a> is a
standardised collection of scRNA-seq datasets and LaminDB makes it easy
to query and access data in this repository. We will go through the
steps of finding and downloading a dataset using
<strong>{laminr}</strong>, performing some simple analysis using
<strong>{Seurat}</strong> and saving the results your own LaminDB
database.</p>
</div>
<div class="section level2">
<h2 id="before-we-start">Before we start<a class="anchor" aria-label="anchor" href="#before-we-start"></a>
</h2>
<p>Before we go begin, please take some time to check out the Getting
Started vignette (<code><a href="../articles/laminr.html">vignette("laminr", package = "laminr")</a></code>).
In particular, make sure you have run the commands in the “Initial
Setup” section.</p>
<p>Once that is done, we can load the <strong>{laminr}</strong>
library.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://laminr.lamin.ai">laminr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="connecting-to-lamindb">Connecting to LaminDB<a class="anchor" aria-label="anchor" href="#connecting-to-lamindb"></a>
</h2>
<p>The first thing we need to do is connect to the LaminDB database. For
this tutorial, we will connect a default instance (where we will store
results) and the CELLxGENE instance that we will search for
datasets.</p>
<div class="section level3">
<h3 id="connect-to-the-default-instance">Connect to the default instance<a class="anchor" aria-label="anchor" href="#connect-to-the-default-instance"></a>
</h3>
<p>We will start by connecting to your default LaminDB instance. You can
set set the default instance using the <code>lamin</code> CLI on the
command line:</p>
<pre class="shell"><code>lamin connect &lt;owner&gt;/&lt;name&gt;</code></pre>
<p>Once a default instance has been set, we can connect to it with
<strong>{laminr}</strong>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/connect.html">connect</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #FFFF55;">!</span> schema module 'bionty' is not installed → no access to its labels &amp; registries (resolve via `pip install bionty`)</span></span>
<span><span class="co">#&gt; <span style="color: #55FF55;">→</span> connected lamindb: laminlabs/cellxgene</span></span>
<span><span class="va">db</span></span>
<span><span class="co">#&gt; <span style="color: #55FF55; font-weight: bold;">cellxgene</span></span></span>
<span><span class="co">#&gt; <span style="color: #FF55FF; font-style: italic;">  Core registries</span></span></span>
<span><span class="co">#&gt;     $Run</span></span>
<span><span class="co">#&gt;     $User</span></span>
<span><span class="co">#&gt;     $Param</span></span>
<span><span class="co">#&gt;     $ULabel</span></span>
<span><span class="co">#&gt;     $Feature</span></span>
<span><span class="co">#&gt;     $Storage</span></span>
<span><span class="co">#&gt;     $Artifact</span></span>
<span><span class="co">#&gt;     $Transform</span></span>
<span><span class="co">#&gt;     $Collection</span></span>
<span><span class="co">#&gt;     $FeatureSet</span></span>
<span><span class="co">#&gt;     $ParamValue</span></span>
<span><span class="co">#&gt;     $FeatureValue</span></span>
<span><span class="co">#&gt; <span style="color: #FF55FF; font-style: italic;">  Additional modules</span></span></span>
<span><span class="co">#&gt;     bionty</span></span></code></pre></div>
<p>This gives us an object we can use to interact with the database.</p>
<p><strong>Note</strong> that only the default instance can create new
records. This tutorial assumes you have access to an instance where you
have permission to add data.</p>
</div>
<div class="section level3">
<h3 id="track-data-provenance">Track data provenance<a class="anchor" aria-label="anchor" href="#track-data-provenance"></a>
</h3>
<p>Before we start, we will track the code that is run in this
notebook.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">db</span><span class="op">$</span><span class="fu">track</span><span class="op">(</span><span class="st">"I8BlHXFXqZOG0000"</span>, path <span class="op">=</span> <span class="st">"example_workflow.Rmd"</span><span class="op">)</span></span></code></pre></div>
<p>Tip: The ID should be obtained by running
<code>db$track(path = "example_workflow.Rmd")</code> and copying the ID
from the output.</p>
</div>
<div class="section level3">
<h3 id="connect-to-the-cellxgene-instance">Connect to the CELLxGENE instance<a class="anchor" aria-label="anchor" href="#connect-to-the-cellxgene-instance"></a>
</h3>
<p>We can connect to other instances by providing a slug to the
<code><a href="../reference/connect.html">connect()</a></code> function. Instances connected to in this way can
be used to query data but cannot make any changes. Let’s connect to the
CELLxGENE instance:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellxgene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/connect.html">connect</a></span><span class="op">(</span><span class="st">"laminlabs/cellxgene"</span><span class="op">)</span></span>
<span><span class="va">cellxgene</span></span>
<span><span class="co">#&gt; <span style="color: #55FF55; font-weight: bold;">cellxgene</span></span></span>
<span><span class="co">#&gt; <span style="color: #FF55FF; font-style: italic;">  Core registries</span></span></span>
<span><span class="co">#&gt;     $Run</span></span>
<span><span class="co">#&gt;     $User</span></span>
<span><span class="co">#&gt;     $Param</span></span>
<span><span class="co">#&gt;     $ULabel</span></span>
<span><span class="co">#&gt;     $Feature</span></span>
<span><span class="co">#&gt;     $Storage</span></span>
<span><span class="co">#&gt;     $Artifact</span></span>
<span><span class="co">#&gt;     $Transform</span></span>
<span><span class="co">#&gt;     $Collection</span></span>
<span><span class="co">#&gt;     $FeatureSet</span></span>
<span><span class="co">#&gt;     $ParamValue</span></span>
<span><span class="co">#&gt;     $FeatureValue</span></span>
<span><span class="co">#&gt; <span style="color: #FF55FF; font-style: italic;">  Additional modules</span></span></span>
<span><span class="co">#&gt;     bionty</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="downloading-a-dataset">Downloading a dataset<a class="anchor" aria-label="anchor" href="#downloading-a-dataset"></a>
</h2>
<p>In Lamin, artifacts are objects that contain information (single-cell
data, images, data frames etc.) as well as associated metadata. You can
see what artifacts are available using the database instance object.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellxgene</span><span class="op">$</span><span class="va">Artifact</span><span class="op">$</span><span class="fu">df</span><span class="op">(</span>limit <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;     id suffix X_accessor n_objects visibility</span></span>
<span><span class="co">#&gt; 1 2846        tiledbsoma       290          1</span></span>
<span><span class="co">#&gt; 2 3665        tiledbsoma       330          1</span></span>
<span><span class="co">#&gt; 3 1270  .h5ad    AnnData        NA          1</span></span>
<span><span class="co">#&gt; 4 2840 .ipynb       &lt;NA&gt;        NA          0</span></span>
<span><span class="co">#&gt; 5 2842  .html       &lt;NA&gt;        NA          0</span></span>
<span><span class="co">#&gt;                                                                      key</span></span>
<span><span class="co">#&gt; 1                                            cell-census/2023-12-15/soma</span></span>
<span><span class="co">#&gt; 2                                            cell-census/2024-07-01/soma</span></span>
<span><span class="co">#&gt; 3 cell-census/2023-07-25/h5ads/7a0a8891-9a22-4549-a55b-c2aca23c3a2a.h5ad</span></span>
<span><span class="co">#&gt; 4                                                                   &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 5                                                                   &lt;NA&gt;</span></span>
<span><span class="co">#&gt;                    uid         size                   hash</span></span>
<span><span class="co">#&gt; 1 FYMewVq5twKMDXVy0000 635848093433 Mfyw8VuqftX5REITfQH_yg</span></span>
<span><span class="co">#&gt; 2 FYMewVq5twKMDXVy0001 870700998221 bzrXBPNvitSVKvb3GG38_w</span></span>
<span><span class="co">#&gt; 3 tczTlSHFPOcAcBnfyxKA   1297573950 UlsVvBz9kMzn2r9RdoAAOg</span></span>
<span><span class="co">#&gt; 4 JIIPyQX5l9qELPl42d75        36297 gNdUkonYgQJP_Mi3xLzt_g</span></span>
<span><span class="co">#&gt; 5 Whyxwf3k2GjJwTPCl1FK       716529 BDGZac3qU3oLVFpO035Qhg</span></span>
<span><span class="co">#&gt;                            description n_observations is_latest X_hash_type</span></span>
<span><span class="co">#&gt; 1                    Census 2023-12-15       68683222     FALSE       md5-d</span></span>
<span><span class="co">#&gt; 2                    Census 2024-07-01      115556140      TRUE       md5-d</span></span>
<span><span class="co">#&gt; 3      Supercluster: Hippocampal CA1-3          74979     FALSE       md5-n</span></span>
<span><span class="co">#&gt; 4 Source of transform G69jtgzKO0eJ6K79             NA     FALSE         md5</span></span>
<span><span class="co">#&gt; 5   Report of run UAAiLAi0BrLvlKnsuvP3             NA     FALSE         md5</span></span>
<span><span class="co">#&gt;      type                       created_at X_key_is_virtual</span></span>
<span><span class="co">#&gt; 1 dataset 2024-07-12T12:12:16.091881+00:00            FALSE</span></span>
<span><span class="co">#&gt; 2 dataset 2024-07-16T12:52:01.424629+00:00            FALSE</span></span>
<span><span class="co">#&gt; 3    &lt;NA&gt; 2023-11-28T21:46:12.685907+00:00            FALSE</span></span>
<span><span class="co">#&gt; 4    &lt;NA&gt; 2024-01-29T08:32:13.311741+00:00             TRUE</span></span>
<span><span class="co">#&gt; 5    &lt;NA&gt; 2024-01-29T08:32:18.346499+00:00             TRUE</span></span>
<span><span class="co">#&gt;                         updated_at    version</span></span>
<span><span class="co">#&gt; 1 2024-09-17T13:00:13.714256+00:00 2023-12-15</span></span>
<span><span class="co">#&gt; 2 2024-09-17T13:01:23.739635+00:00 2024-07-01</span></span>
<span><span class="co">#&gt; 3 2024-01-24T07:10:21.725547+00:00 2023-07-25</span></span>
<span><span class="co">#&gt; 4 2024-01-29T08:32:13.311792+00:00          0</span></span>
<span><span class="co">#&gt; 5 2024-01-30T09:12:06.027928+00:00          1</span></span></code></pre></div>
<p>This is useful, but it’s not the nicest or easiest way to find a
particular dataset. Instead, we will use the Lamin Hub website to find
the data we want to load.</p>
<ol style="list-style-type: decimal">
<li>Open a browser and go to <a href="https://lamin.ai/laminlabs/cellxgene" class="external-link uri">https://lamin.ai/laminlabs/cellxgene</a>
</li>
<li>On the top toolbar, click the “Artifacts” tab</li>
<li>Use the search field and the filters to find a dataset you are
interested in.</li>
</ol>
<ul>
<li>We use the “Suffix” filter to find <code>.h5ad</code> files and
search for “renal cell carcinoma”</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>Select the entry for the dataset you want to load to open a page
with more details</li>
<li>Click the copy button at the top right, this copies a command
including the ID for the artifact</li>
</ol>
<p>Once we have the artifact ID, we can load information about the
artifact, similar to what we see on the website. Notice that we use a
slightly different command to what we copied from the website.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">artifact</span> <span class="op">&lt;-</span> <span class="va">cellxgene</span><span class="op">$</span><span class="va">Artifact</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"7dVluLROpalzEh8mNyxk"</span><span class="op">)</span></span>
<span><span class="va">artifact</span></span>
<span><span class="co">#&gt; <span style="color: #55FF55; font-weight: bold;">Artifact</span>(uid='7dVluLROpalzEh8mNyxk', description='Renal cell carcinoma, pre aPD1, kidney Puck_200727_12', key='cell-census/2023-12-15/h5ads/02faf712-92d4-4589-bec7-13105059cf86.h5ad', id=1742, run_id=22, hash='YNYuokfAoDFxdaRILjmU9w', size=13997860, suffix='.h5ad', storage_id=2, version='2023-12-15', _accessor='AnnData', is_latest=TRUE, transform_id=16, _hash_type='md5-n', created_at='2024-01-11T09:13:23.143694+00:00', created_by_id=1, updated_at='2024-01-24T07:17:47.009288+00:00', visibility=1, n_observations=17612, _key_is_virtual=FALSE)</span></span></code></pre></div>
<p>So far we have only retrieved the metadata about this object. To
download the data itself we need to run another command.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="va">artifact</span><span class="op">$</span><span class="fu">load</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |                                                                      |   1%  |                                                                              |=                                                                     |   1%  |                                                                              |=                                                                     |   2%  |                                                                              |==                                                                    |   2%  |                                                                              |==                                                                    |   3%  |                                                                              |===                                                                   |   4%  |                                                                              |===                                                                   |   5%  |                                                                              |====                                                                  |   5%  |                                                                              |====                                                                  |   6%  |                                                                              |=====                                                                 |   6%  |                                                                              |=====                                                                 |   7%  |                                                                              |=====                                                                 |   8%  |                                                                              |======                                                                |   8%  |                                                                              |======                                                                |   9%  |                                                                              |=======                                                               |   9%  |                                                                              |=======                                                               |  10%  |                                                                              |=======                                                               |  11%  |                                                                              |========                                                              |  11%  |                                                                              |========                                                              |  12%  |                                                                              |=========                                                             |  12%  |                                                                              |=========                                                             |  13%  |                                                                              |==========                                                            |  14%  |                                                                              |==========                                                            |  15%  |                                                                              |===========                                                           |  15%  |                                                                              |===========                                                           |  16%  |                                                                              |============                                                          |  17%  |                                                                              |============                                                          |  18%  |                                                                              |=============                                                         |  18%  |                                                                              |=============                                                         |  19%  |                                                                              |==============                                                        |  19%  |                                                                              |==============                                                        |  20%  |                                                                              |==============                                                        |  21%  |                                                                              |===============                                                       |  21%  |                                                                              |===============                                                       |  22%  |                                                                              |================                                                      |  22%  |                                                                              |================                                                      |  23%  |                                                                              |================                                                      |  24%  |                                                                              |=================                                                     |  24%  |                                                                              |=================                                                     |  25%  |                                                                              |==================                                                    |  25%  |                                                                              |==================                                                    |  26%  |                                                                              |===================                                                   |  26%  |                                                                              |===================                                                   |  27%  |                                                                              |===================                                                   |  28%  |                                                                              |====================                                                  |  28%  |                                                                              |====================                                                  |  29%  |                                                                              |=====================                                                 |  29%  |                                                                              |=====================                                                 |  30%  |                                                                              |=====================                                                 |  31%  |                                                                              |======================                                                |  31%  |                                                                              |======================                                                |  32%  |                                                                              |=======================                                               |  32%  |                                                                              |=======================                                               |  33%  |                                                                              |========================                                              |  34%  |                                                                              |========================                                              |  35%  |                                                                              |=========================                                             |  35%  |                                                                              |=========================                                             |  36%  |                                                                              |==========================                                            |  37%  |                                                                              |==========================                                            |  38%  |                                                                              |===========================                                           |  38%  |                                                                              |===========================                                           |  39%  |                                                                              |============================                                          |  39%  |                                                                              |============================                                          |  40%  |                                                                              |============================                                          |  41%  |                                                                              |=============================                                         |  41%  |                                                                              |=============================                                         |  42%  |                                                                              |==============================                                        |  42%  |                                                                              |==============================                                        |  43%  |                                                                              |==============================                                        |  44%  |                                                                              |===============================                                       |  44%  |                                                                              |===============================                                       |  45%  |                                                                              |================================                                      |  45%  |                                                                              |================================                                      |  46%  |                                                                              |=================================                                     |  46%  |                                                                              |=================================                                     |  47%  |                                                                              |=================================                                     |  48%  |                                                                              |==================================                                    |  48%  |                                                                              |==================================                                    |  49%  |                                                                              |===================================                                   |  49%  |                                                                              |===================================                                   |  50%  |                                                                              |===================================                                   |  51%  |                                                                              |====================================                                  |  51%  |                                                                              |====================================                                  |  52%  |                                                                              |=====================================                                 |  52%  |                                                                              |=====================================                                 |  53%  |                                                                              |=====================================                                 |  54%  |                                                                              |======================================                                |  54%  |                                                                              |======================================                                |  55%  |                                                                              |=======================================                               |  55%  |                                                                              |=======================================                               |  56%  |                                                                              |========================================                              |  57%  |                                                                              |========================================                              |  58%  |                                                                              |=========================================                             |  58%  |                                                                              |=========================================                             |  59%  |                                                                              |==========================================                            |  59%  |                                                                              |==========================================                            |  60%  |                                                                              |==========================================                            |  61%  |                                                                              |===========================================                           |  61%  |                                                                              |===========================================                           |  62%  |                                                                              |============================================                          |  62%  |                                                                              |============================================                          |  63%  |                                                                              |============================================                          |  64%  |                                                                              |=============================================                         |  64%  |                                                                              |=============================================                         |  65%  |                                                                              |==============================================                        |  65%  |                                                                              |==============================================                        |  66%  |                                                                              |===============================================                       |  66%  |                                                                              |===============================================                       |  67%  |                                                                              |===============================================                       |  68%  |                                                                              |================================================                      |  68%  |                                                                              |================================================                      |  69%  |                                                                              |=================================================                     |  69%  |                                                                              |=================================================                     |  70%  |                                                                              |=================================================                     |  71%  |                                                                              |==================================================                    |  71%  |                                                                              |==================================================                    |  72%  |                                                                              |===================================================                   |  72%  |                                                                              |===================================================                   |  73%  |                                                                              |===================================================                   |  74%  |                                                                              |====================================================                  |  74%  |                                                                              |====================================================                  |  75%  |                                                                              |=====================================================                 |  75%  |                                                                              |=====================================================                 |  76%  |                                                                              |======================================================                |  76%  |                                                                              |======================================================                |  77%  |                                                                              |======================================================                |  78%  |                                                                              |=======================================================               |  78%  |                                                                              |=======================================================               |  79%  |                                                                              |========================================================              |  79%  |                                                                              |========================================================              |  80%  |                                                                              |========================================================              |  81%  |                                                                              |=========================================================             |  81%  |                                                                              |=========================================================             |  82%  |                                                                              |==========================================================            |  82%  |                                                                              |==========================================================            |  83%  |                                                                              |===========================================================           |  84%  |                                                                              |===========================================================           |  85%  |                                                                              |============================================================          |  85%  |                                                                              |============================================================          |  86%  |                                                                              |=============================================================         |  87%  |                                                                              |=============================================================         |  88%  |                                                                              |==============================================================        |  88%  |                                                                              |==============================================================        |  89%  |                                                                              |===============================================================       |  89%  |                                                                              |===============================================================       |  90%  |                                                                              |===============================================================       |  91%  |                                                                              |================================================================      |  91%  |                                                                              |================================================================      |  92%  |                                                                              |=================================================================     |  92%  |                                                                              |=================================================================     |  93%  |                                                                              |=================================================================     |  94%  |                                                                              |==================================================================    |  94%  |                                                                              |==================================================================    |  95%  |                                                                              |===================================================================   |  95%  |                                                                              |===================================================================   |  96%  |                                                                              |====================================================================  |  96%  |                                                                              |====================================================================  |  97%  |                                                                              |====================================================================  |  98%  |                                                                              |===================================================================== |  98%  |                                                                              |===================================================================== |  99%  |                                                                              |======================================================================|  99%  |                                                                              |======================================================================| 100%</span></span>
<span><span class="va">adata</span></span>
<span><span class="co">#&gt; AnnData object with n_obs × n_vars = 17612 × 23254</span></span>
<span><span class="co">#&gt;     obs: 'n_genes', 'n_UMIs', 'log10_n_UMIs', 'log10_n_genes', 'Cell_Type', 'cell_type_ontology_term_id', 'organism_ontology_term_id', 'tissue_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'donor_id', 'is_primary_data', 'suspension_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage'</span></span>
<span><span class="co">#&gt;     var: 'gene', 'n_beads', 'n_UMIs', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype'</span></span>
<span><span class="co">#&gt;     uns: 'Cell_Type_colors', 'schema_version', 'title'</span></span>
<span><span class="co">#&gt;     obsm: 'X_spatial'</span></span></code></pre></div>
<p>This dataset has been stored as an <a href="https://anndata.readthedocs.io" class="external-link"><code>AnnData</code></a> object.
In the next sections we will convert it to a <a href="https://satijalab.org/seurat/" class="external-link"><code>Seurat</code></a> object and
perform some simple analysis.</p>
</div>
<div class="section level2">
<h2 id="convert-to-seurat">Convert to Seurat<a class="anchor" aria-label="anchor" href="#convert-to-seurat"></a>
</h2>
<p>There are various approaches for converting between different
single-cell objects, some of which are described in the <a href="https://www.sc-best-practices.org/introduction/interoperability.html" class="external-link">Interoperability
chapter</a> of the Single-cell Best Practices book.</p>
<p>Because we already have the data loaded in memory, the simplest
option is to extract the information we need and create a new
<code>Seurat</code> object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">adata</span><span class="op">$</span><span class="va">X</span><span class="op">)</span>,</span>
<span>  meta.data <span class="op">=</span> <span class="va">adata</span><span class="op">$</span><span class="va">obs</span>,</span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Data is of class dgRMatrix. Coercing to dgCMatrix.</span></span>
<span><span class="va">seurat</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 23254 features across 17612 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (23254 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>We could perform any normal analysis using <strong>{Seurat}</strong>
but as an example we will calculate marker genes for each of the
annotated cell types. To make things a bit quicker we only test the
first 1000 genes but if you have a few minutes you can get results for
all features.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set cell identities to the provided cell type annotation</span></span>
<span><span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Cell_Type"</span></span>
<span><span class="co"># Normalise the data</span></span>
<span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">)</span></span>
<span><span class="co">#&gt; Normalizing layer: counts</span></span>
<span><span class="co"># Test for marker genes</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat</span>,</span>
<span>  features <span class="op">=</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Cells.html" class="external-link">Features</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating cluster Epithelial</span></span>
<span><span class="co">#&gt; Calculating cluster Fibroblast</span></span>
<span><span class="co">#&gt; For a (much!) faster implementation of the Wilcoxon Rank Sum Test,</span></span>
<span><span class="co">#&gt; (default method for FindMarkers) please install the presto package</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; install.packages('devtools')</span></span>
<span><span class="co">#&gt; devtools::install_github('immunogenomics/presto')</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; After installation of presto, Seurat will automatically use the more </span></span>
<span><span class="co">#&gt; efficient implementation (no further action necessary).</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="co">#&gt; Calculating cluster Myeloid</span></span>
<span><span class="co">#&gt; Calculating cluster Tumor</span></span>
<span><span class="co">#&gt; Warning: The following tests were not performed:</span></span>
<span><span class="co">#&gt; Warning: When testing Epithelial versus all:</span></span>
<span><span class="co">#&gt;  Cell group 1 has fewer than 3 cells</span></span>
<span><span class="co"># The output is a data.frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span>
<span><span class="co">#&gt;                        p_val avg_log2FC pct.1 pct.2    p_val_adj    cluster</span></span>
<span><span class="co">#&gt; ENSG00000164283 1.030703e-89  2.7485040 0.205 0.048 2.396797e-85 Fibroblast</span></span>
<span><span class="co">#&gt; ENSG00000116016 3.606838e-38  2.0721038 0.152 0.051 8.387340e-34 Fibroblast</span></span>
<span><span class="co">#&gt; ENSG00000074800 5.097282e-25 -0.9810317 0.185 0.366 1.185322e-20 Fibroblast</span></span>
<span><span class="co">#&gt; ENSG00000112715 6.663398e-18 -1.1826785 0.078 0.202 1.549507e-13 Fibroblast</span></span>
<span><span class="co">#&gt; ENSG00000140416 1.844156e-17 -0.6994000 0.175 0.326 4.288400e-13 Fibroblast</span></span>
<span><span class="co">#&gt; ENSG00000125810 8.916133e-15  1.8102270 0.057 0.019 2.073358e-10 Fibroblast</span></span>
<span><span class="co">#&gt;                            gene</span></span>
<span><span class="co">#&gt; ENSG00000164283 ENSG00000164283</span></span>
<span><span class="co">#&gt; ENSG00000116016 ENSG00000116016</span></span>
<span><span class="co">#&gt; ENSG00000074800 ENSG00000074800</span></span>
<span><span class="co">#&gt; ENSG00000112715 ENSG00000112715</span></span>
<span><span class="co">#&gt; ENSG00000140416 ENSG00000140416</span></span>
<span><span class="co">#&gt; ENSG00000125810 ENSG00000125810</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="store-the-results-in-lamindb">Store the results in LaminDB<a class="anchor" aria-label="anchor" href="#store-the-results-in-lamindb"></a>
</h2>
<p>Now that we have our results, we can save them to the LaminDB
instance.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">seurat</span>, <span class="va">seu_path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">db</span><span class="op">$</span><span class="va">Artifact</span><span class="op">$</span><span class="fu">from_df</span><span class="op">(</span></span>
<span>  <span class="va">markers</span>,</span>
<span>  description <span class="op">=</span> <span class="st">"Marker genes for renal cell carcinoma dataset"</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="fu">save</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">db</span><span class="op">$</span><span class="va">Artifact</span><span class="op">$</span><span class="fu">from_path</span><span class="op">(</span></span>
<span>  <span class="va">seu_path</span>,</span>
<span>  description <span class="op">=</span> <span class="st">"Seurat object for renal cell carcinoma dataset"</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="fu">save</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="close-the-connection">Close the connection<a class="anchor" aria-label="anchor" href="#close-the-connection"></a>
</h2>
<p>Finally, we can close the connection to the database.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">db</span><span class="op">$</span><span class="fu">finish</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="render-and-upload-the-notebook">Render and upload the notebook<a class="anchor" aria-label="anchor" href="#render-and-upload-the-notebook"></a>
</h2>
<p>You can render this notebook to HTML:</p>
<ul>
<li><p>In RStudio, click the “Knit” button</p></li>
<li>
<p>From the command line, run:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">'rmarkdown::render("example_workflow.Rmd")'</span></span></code></pre></div>
</li>
<li>
<p>Or use the <code>rmarkdown</code> package in R:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"example_workflow.Rmd"</span><span class="op">)</span></span></code></pre></div>
</li>
</ul>
<p>And then save it to your LaminDB instance using the
<code>lamin</code> CLI:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">lamin</span> save example_workflow.Rmd</span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://cannoodt.dev" class="external-link">Robrecht Cannoodt</a>, <a href="https://lazappi.id.au" class="external-link">Luke Zappia</a>, <a href="https://data-intuitive.com" class="external-link">Data Intuitive</a>, <a href="https://lamin.ai" class="external-link">Lamin Labs</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
